### ISSUE TITLE
Loss curves

### ISSUE THREAD

**Issue by omair-kg** [mrjel/noise-as-targets-tensorflow#2]
Hi, This isn't really an issue as such, more of a request. Could you possible share your loss curves. How the MSE loss varies with epoch number. Thanks in anticipation.

**Comment by janericlenssen** [comment]
Hi,
shown below is the MSE loss curve I got with the current version of the script.
No epochs on x-axis but I have a training set of 50000 images (cifar10 training set) and a batch_size of 256 so 195 steps are approximately one epoch.

![natenc-loss](https://user-images.githubusercontent.com/9112195/28025641-4329ddf0-6594-11e7-8777-9bbe580f33e3.png)

What I observe is that loss is dropping very fast during the reassignment epochs and slower in other epochs, if at all.
It indicates that the reassignment is working better than the actual encoder training.

**Comment by omair-kg** [comment]
I'll just comment here instead of opening another issue. I noticed in your read me file that you 'Optimizes the assignment of target vectors every 3 epochs using the hungarian algorithm'. 

In the paper though this is done in each batch before the backward pass (see Algo 1 in paper). Is this a typo or was this something you noticed for better optimization?

Edit:
---------------------------------------
So i rechecked the code and the reassignment is done every 3rd epoch. Did you also try it out for the scenario described in the paper (per batch: forward pass, reassignment, backward pass, update)?

**Comment by omair-kg** [comment]
"What I observe is that loss is dropping very fast during the reassignment epochs and slower in other epochs, if at all.
It indicates that the reassignment is working better than the actual encoder training."

Similar observation for me if i do the reassignment every 3rd epoch. With a reassignment in every batch as proposed in the paper, the training is a bit unstable and collapses more readily.

**Comment by janericlenssen** [comment]
Hi,
From page 5 of the paper in the optimization details: "Unless mentioned otherwise, we permute
the  assignments  within  batches  every 3 epochs". That is where i got the 3 epoch interval from.

What i do at the moment:
In every third epoch, for each batch:
1. Compute Forward pass for the batch
2. Reassign the target vectors for the batch
3. Compute loss using the reassigned target vectors and network output (here it does not matter if you reuse the network output from step 1 or compute it again since you did not change the weights. Reusing is faster obviously but just computing it again let me sess.run() the train op which was more straight forward to implement.)
4. Backward step
5. Parameter update

As far as i understand it, it is exactly like they do it in the paper, or am i wrong?
Maybe my short description in the readme is misleading.

**Comment by omair-kg** [comment]
Ok, so as i understood it from the paper and the [inference blog post](http://www.inference.vc/unsupervised-learning-by-predicting-noise-an-information-maximization-view-2/), on every 3rd epoch the instances are reassigned to new mini-batches. E.g. if an instance x_i belongs to mini batch y_i, then on the third epoch the mini batches are reshuffled and now instance x_i belongs to some other mini batch z_i. This ensures that problem of getting stuck in a local minima as highlighted in the blog post is mitigated.

The hungarian reassignment for mini batches is supposed to be done IMO in every epoch and the whole training list needs to be shuffled on every 3rd epoch.

**Comment by janericlenssen** [comment]
I do it exactly the other way around:
- Shuffle training set (for new combination of batches) after every epoch
- Perform the hungarian reassignment within batches every 3rd epoch

I am also still convinced that this is the way the authors intended:
"we **permute** the assignments **within** batches every 3 epochs" seems to be quite clear. But well, only the authors will know for sure. ;)

Anyway, it is interesting to see if you get better results with your approach.

**Comment by omair-kg** [comment]
I'm basing my understanding on Sec. 3.2 SGD and Algo 1. Overall though, I don't think the order of shuffling and hungarian reassignment will make too much difference. So far my results correspond to yours but with less number of epochs. Training also seems to be a bit unstable with the model collapsing after ~30 epochs.

**Comment by omair-kg** [comment]
I do believe you understood it correctly. The reassignment is done on every 3rd epoch. The 'Effects of permutations' section cleared that up for me.
